### PR TITLE
LibWeb: Implement MessagePort.postMessage closer to the spec

### PR DESCRIPTION
**Pull request by ADKaster**
Use a LocalSocket to represent the connection between two message ports.

The concept of the port message queue is still missing, however. When
that concept is implemented, the "steps" in step 7 of the message port
transfer steps will need to send the serialized data over the connected
socketpair and run in the event loop of the process that holds onto the
other side of the message port. Doing this should allow centralizing the
behavior of postMessage for Window, MessagePorts and Workers.

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/VM.h [hunk 1]
```diff
@@ -114,8 +114,16 @@ class VM : public RefCounted<VM> {
     // https://tc39.es/ecma262/#running-execution-context
     // At any point in time, there is at most one execution context per agent that is actually executing code.
     // This is known as the agent's running execution context.
-    ExecutionContext& running_execution_context() { return *m_execution_context_stack.last(); }
-    ExecutionContext const& running_execution_context() const { return *m_execution_context_stack.last(); }
+    ExecutionContext& running_execution_context()
+    {
+        VERIFY(!m_execution_context_stack.is_empty());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/VM.h:119]: Big fan of this change, and the following one! Just wondering though, just given how often this happens - for people who aren't familiar with the issue, I wonder if it's also worth us leaving a comment / mention the workaround that we often need to make for this in LibWeb-land

File: Userland/Libraries/LibWeb/HTML/MessagePort.cpp [hunk 2]
```diff
@@ -103,50 +124,124 @@ void MessagePort::entangle_with(MessagePort& remote_port)
...
-    // https://html.spec.whatwg.org/multipage/web-messaging.html#message-port-post-message-steps
+// https://html.spec.whatwg.org/multipage/web-messaging.html#dom-messageport-postmessage
+WebIDL::ExceptionOr<void> MessagePort::post_message(JS::Value message, StructuredSerializeOptions const& options)
+{
+    // 1. Let targetPort be the port with which this MessagePort is entangled, if any; otherwise let it be null.
+    JS::GCPtr<MessagePort> target_port = m_remote_port;
 
-    // FIXME: 1. Let transfer be options["transfer"].
+    // 2. Run the message port post message steps providing targetPort, message and options.
+    return TRY(message_port_post_message_steps(target_port, message, options));
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/MessagePort.cpp:157]:
```suggestion
    return message_port_post_message_steps(target_port, message, options);
```

File: Userland/Libraries/LibWeb/HTML/MessagePort.cpp [hunk 3]
```diff
@@ -103,50 +124,124 @@ void MessagePort::entangle_with(MessagePort& remote_port)
...
-    auto* target_port = m_remote_port.ptr();
+    JS::GCPtr<MessagePort> target_port = m_remote_port;
 
-    // FIXME: 2. Let options be «[ "transfer" → transfer ]».
+    // 2. Let options be «[ "transfer" → transfer ]».
+    auto options = StructuredSerializeOptions { transfer };
 
-    // 3. Run the message port post message steps providing targetPort, message and options.
+    // 3. Run the message port post message steps providing this, targetPort, message and options.
+    return TRY(message_port_post_message_steps(target_port, message, options));
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/MessagePort.cpp:147]:
```suggestion
    return message_port_post_message_steps(target_port, message, options);
```

File: Userland/Libraries/LibJS/Runtime/VM.h (hunk 1 above)
Comment (ADKaster) [reply Userland/Libraries/LibJS/Runtime/VM.h:119]: I don't think this needs a comment for the workaround here in LibJS. It feels like bad layering, and we are familiar enough with the issue that a quick "hey what's this execution context stack thing about" in #browser will yield the right answer.

**Pull request merged**